[dynamo] Guard serialization for FUNCTORCH_STACK_MATCH

zhxchen17 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #152723
* #152721
* #152716
* #152687
* __->__ #152616
* #152615

Make Functorch interpreters serializable most of the time, so that we can save the guards on functorch states.

## Test Cases:

0. torch.compile() without functorch layers present. Guard should fail with any layer being pushed.
1. torch.compile() nested in vmap.
2. torch.compile() nested in grad.
3. torch.compile() nested in jvp + vmap
4. torch.compile() nested functionalize
5. torch.compile() nested in vmap + grad

Differential Revision: [D74008787](https://our.internmc.facebook.com/intern/diff/D74008787/)

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @jerryzh168 @voznesenskym @penguinwu @EikanWang @Guobing-Chen @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames

zhxchen17 committed:
[dynamo] Guard serialization for FUNCTORCH_STACK_MATCH

Make Functorch interpreters serializable most of the time, so that we can save the guards on functorch states.

Differential Revision: [D74008787](https://our.internmc.facebook.com/intern/diff/D74008787/)

[ghstack-poisoned]

zhxchen17 committed:
Update on "[dynamo] Guard serialization for FUNCTORCH_STACK_MATCH"


Make Functorch interpreters serializable most of the time, so that we can save the guards on functorch states.

Differential Revision: [D74008787](https://our.internmc.facebook.com/intern/diff/D74008787/)

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

jansel reviewed (changes requested):
Add tests?

zhxchen17 committed:
Update on "[dynamo] Guard serialization for FUNCTORCH_STACK_MATCH"


Make Functorch interpreters serializable most of the time, so that we can save the guards on functorch states.

Differential Revision: [D74008787](https://our.internmc.facebook.com/intern/diff/D74008787/)

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

zou3519 commented on aten/src/ATen/functorch/DynamicLayer.h:
```diff
@@ -15,6 +15,8 @@
 #include <ATen/functorch/ADInterpreters.h>
 #include <ATen/functorch/FunctionalizeInterpreter.h>
 
+#include <nlohmann/json.hpp>
```
Do you need to serialize this in C++?

zhxchen17 replied:
Yeah I don't need this header in this file anymore (leftover from the previous version).

zhxchen17 committed:
Update on "[dynamo] Guard serialization for FUNCTORCH_STACK_MATCH"


Make Functorch interpreters serializable most of the time, so that we can save the guards on functorch states.

Differential Revision: [D74008787](https://our.internmc.facebook.com/intern/diff/D74008787/)

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

zhxchen17 committed:
Update on "[dynamo] Guard serialization for FUNCTORCH_STACK_MATCH"


Make Functorch interpreters serializable most of the time, so that we can save the guards on functorch states.

Differential Revision: [D74008787](https://our.internmc.facebook.com/intern/diff/D74008787/)

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 jerryzh168 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

zhxchen17 commented:
Added more tests.

zhxchen17 committed:
Update on "[dynamo] Guard serialization for FUNCTORCH_STACK_MATCH"


Make Functorch interpreters serializable most of the time, so that we can save the guards on functorch states.

## Test Cases:

0. torch.compile() without functorch layers present. Guard should fail with any layer being pushed.
1. torch.compile() nested in vmap.
2. torch.compile() nested in grad.
3. torch.compile() nested in jvp + vmap
4. torch.compile() nested functionalize
5. torch.compile() nested in vmap + grad

Differential Revision: [D74008787](https://our.internmc.facebook.com/intern/diff/D74008787/)

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 jerryzh168 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]
